lib/psabpf_meter.c

```diff
@@ -173,111 +139,283 @@ int psabpf_meter_entry_data(psabpf_meter_entry_t *entry,
+    if (return_code != 0) {
+        return_code = errno;
+        fprintf(stderr, "failed to get entry: %s\n", strerror(return_code));
+        goto clean_up;
+    }
+
+    psabpf_meter_data_t data;
+    memcpy(&data, value_buffer, sizeof(data));
+    return_code = convert_meter_data_to_entry(&data, &ctx->current_entry);
+    if (return_code != NO_ERROR)
+        goto clean_up;
+
+    ret_instance = &ctx->current_entry;
+
+clean_up:
+    if (next_key != NULL)
+        free(next_key);
+    if (value_buffer == NULL)
+        free(value_buffer);
+
+    return ret_instance;
+}
+
+int psabpf_meter_entry_update(psabpf_meter_ctx_t *ctx, psabpf_meter_entry_t *entry) {
     int return_code = NO_ERROR;
     uint64_t bpf_flags = BPF_F_LOCK;
     char *value_buffer = NULL;
-    char *index_buffer = NULL;
     psabpf_meter_data_t data;
 
-    return_code = check_index(ctx, entry);
-    if (return_code != NO_ERROR)
-        return return_code;
+    if (ctx == NULL || entry == NULL)
+        return EINVAL;
+
+    if (entry->index_sfs.n_fields == 0) {
+        fprintf(stderr, "Index not provided");
+        return EINVAL;
+    }
 
     return_code = convert_meter_entry_to_data(entry, &data);
     if (return_code != NO_ERROR)
         return return_code;
 
-    index_buffer = malloc(ctx->index_size);
-    value_buffer = malloc(ctx->value_size);
-    if (index_buffer == NULL || value_buffer == NULL) {
+    if (entry->raw_index == NULL)
+        entry->raw_index = malloc(ctx->meter.key_size);
+    value_buffer = malloc(ctx->meter.value_size);
+    if (entry->raw_index == NULL || value_buffer == NULL) {
         fprintf(stderr, "not enough memory\n");
         return_code = ENOMEM;
         goto clean_up;
     }
 
-    memset(value_buffer, 0, ctx->value_size);
+    memset(value_buffer, 0, ctx->meter.value_size);
     memcpy(value_buffer, &data, sizeof(data));
-    memset(index_buffer, 0, ctx->index_size);
-    memcpy(index_buffer, entry->index, entry->index_size);
+    return_code = construct_struct_from_fields(&entry->index_sfs, &ctx->index_fds, entry->raw_index, ctx->meter.key_size);
+    if (return_code != NO_ERROR)
+        goto clean_up;
 
-    return_code = bpf_map_update_elem(ctx->table_fd, index_buffer, value_buffer, bpf_flags);
-    if (return_code != NO_ERROR) {
+    return_code = bpf_map_update_elem(ctx->meter.fd, entry->raw_index, value_buffer, bpf_flags);
+    if (return_code != 0) {
         return_code = errno;
         fprintf(stderr, "failed to set up meter: %s\n", strerror(errno));
         goto clean_up;
     }
 
 clean_up:
     free(value_buffer);
-    free(index_buffer);
     return return_code;
 }
 
-int psabpf_meter_ctx_reset(psabpf_meter_ctx_t *ctx, psabpf_meter_entry_t *entry) {
-    int return_code = psabpf_meter_entry_data(entry, 0, 0, 0, 0);
+int psabpf_meter_entry_reset(psabpf_meter_ctx_t *ctx, psabpf_meter_entry_t *entry) {
+    if (ctx == NULL || entry == NULL)
+        return EINVAL;
+
+    if (entry->index_sfs.n_fields < 1)
```

Comment:
Why number of fields indicates deleting all map entries?

Reply:
When client doesn't provide index, then number of fields is equal to 0. In such case all entries are removed because there is no index.